### DISCUSSION TITLE
Messaging iOS

### DISCUSSION THREAD

**Discussion by alexookah** [capawesome-team/capacitor-firebase#254] [Q&A]
I see this logs in my iOS setup project.

```
- [FirebaseCore][I-COR000003] The default Firebase app has not yet been configured. Add `FirebaseApp.configure()` to your application initialization. This can be done in in the App Delegate's application(_:didFinishLaunchingWithOptions:)` (or the `@main` struct's initializer in SwiftUI). Read more: https://goo.gl/ctyzm8.

- [FirebaseMessaging][I-FCM001000] FIRMessaging Remote Notifications proxy enabled, will swizzle remote notification receiver handlers. If you'd prefer to manually integrate Firebase Messaging, add "FirebaseAppDelegateProxyEnabled" to your Info.plist, and set it to NO. Follow the instructions at:

- [FirebaseMessaging][I-FCM002022] APNS device token not set before retrieving FCM Token for Sender ID '640168526589'. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.

```

I guess about FirebaseApp.configure() there is nothing we should do. right?

I saw also in the [documentation](https://github.com/capawesome-team/capacitor-firebase/tree/main/packages/messaging#ios) of the setup that in info.plist this should be added `FirebaseAppDelegateProxyEnabled`.
Why?

**Comment by alexookah** [comment]
And another question is about this:


```
- ⚡️  TO JS {"notification":{"data":{"com.google.firebase.auth":{"warning":"This fake notification should be forwarded to Firebase Auth."}}}}

- ⚡️  [log] - notificationReceived {"event":{"notification":{"data":{"com.google.firebase.auth":{"warning":"This fake notification should be forwarded to Firebase Auth."}}}}}

- ⚡️  [log] - {"com.google.firebase.auth":{"warning":"This fake notification should be forwarded to Firebase Auth."}}
```

I am getting those fake notifications in my UI. 
Is there something that should be done within the library to skip those?

**Reply by robingenz** [reply]
First time I see this error message. Maybe this helps you: https://stackoverflow.com/a/62405030/6731412

**Reply by alexookah** [reply]
This notification is delivered without FirebaseAppDelegateProxyEnabled and with phone authentication only.
I think firebase sends this to probably to determine that the device trying to authenticate is real without needing to show the recaptcha.

**Comment by robingenz** [comment]
> I see this logs in my iOS setup project.
> 
> 
> 
> ```
> 
> - [FirebaseCore][I-COR000003] The default Firebase app has not yet been configured. Add `FirebaseApp.configure()` to your application initialization. This can be done in in the App Delegate's application(_:didFinishLaunchingWithOptions:)` (or the `@main` struct's initializer in SwiftUI). Read more: https://goo.gl/ctyzm8.
> 
> 
> 
> - [FirebaseMessaging][I-FCM001000] FIRMessaging Remote Notifications proxy enabled, will swizzle remote notification receiver handlers. If you'd prefer to manually integrate Firebase Messaging, add "FirebaseAppDelegateProxyEnabled" to your Info.plist, and set it to NO. Follow the instructions at:
> 
> 
> 
> - [FirebaseMessaging][I-FCM002022] APNS device token not set before retrieving FCM Token for Sender ID '640168526589'. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.
> 
> 
> 
> ```
> 
> 
> 
> I guess about FirebaseApp.configure() there is nothing we should do. right?
> 
> 
> 
> I saw also in the [documentation](https://github.com/capawesome-team/capacitor-firebase/tree/main/packages/messaging#ios) of the setup that in info.plist this should be added `FirebaseAppDelegateProxyEnabled`.
> 
> Why?
> 
> 

Right, you can ignore that.

FirebaseAppDelegateProxyEnabled is necessary to disable swizzling (read more [here](https://firebase.google.com/docs/cloud-messaging/ios/receive?hl=en#handle_messages_with_method_swizzling_disabled)).

**Reply by alexookah** [reply]
once I define the `FirebaseAppDelegateProxyEnabled` for messaging to disable swizzling I get these errors while trying to phone authenticate with iOS.

`ERROR MESSAGE:  {"message":"If app delegate swizzling is disabled, remote notifications received by UIApplicationDelegate need to be forwarded to FIRAuth's canHandleNotificaton: method.","errorMessage":"If app delegate swizzling is disabled, remote notifications received by UIApplicationDelegate need to be forwarded to FIRAuth's canHandleNotificaton: method."}`

Are we sure that an app with phone login and messaging with swizzle disabled, the messaging works correctly?

**Reply by alexookah** [reply]
Maybe this has to be done?
https://firebase.google.com/docs/cloud-messaging/ios/receive?hl=en#handle_messages_with_method_swizzling_disabled

**Reply by alexookah** [reply]
> FirebaseAppDelegateProxyEnabled is necessary to disable swizzling (read more [here](https://firebase.google.com/docs/cloud-messaging/ios/receive?hl=en#handle_messages_with_method_swizzling_disabled)).

I am sorry but as far as I undestand this documentation link does not say that it's required to disable the swizzling.
I understand that by default it is enabled and if someone does not like it or is having issues, it can be disabled.

If it is disabled then the documentation instructions should be followed how to properly use messaging with `swizzling disabled`.

**Reply by robingenz** [reply]
> Maybe this has to be done? [firebase.google.com/docs/cloud-messaging/ios/receive?hl=en#handle_messages_with_method_swizzling_disabled](https://firebase.google.com/docs/cloud-messaging/ios/receive?hl=en#handle_messages_with_method_swizzling_disabled)

This is already done:
https://github.com/capawesome-team/capacitor-firebase/blob/fadf36b317207d54949828b0c6cf361de253f17e/packages/messaging/ios/Plugin/FirebaseMessagingPlugin.swift#L185
and
https://github.com/capawesome-team/capacitor-firebase/blob/fadf36b317207d54949828b0c6cf361de253f17e/packages/messaging/ios/Plugin/FirebaseMessaging.swift#L104

**Reply by robingenz** [reply]
> > FirebaseAppDelegateProxyEnabled is necessary to disable swizzling (read more [here](https://firebase.google.com/docs/cloud-messaging/ios/receive?hl=en#handle_messages_with_method_swizzling_disabled)).
> 
> I am sorry but as far as I undestand this documentation link does not say that it's required to disable the swizzling. I understand that by default it is enabled and if someone does not like it or is having issues, it can be disabled.

Unfortunately, I could not find any official documentation for `FirebaseAppDelegateProxyEnabled`.
Maybe this SO answer helps you: https://stackoverflow.com/a/69077431/6731412
If I remember correctly, then with `FirebaseAppDelegateProxyEnabled` the listener of the plugin (`notificationActionPerformed`, ...) no longer works, because Firebase Messaging takes care of everything itself.

> If it is disabled then the documentation instructions should be followed how to properly use messaging with `swizzling disabled`.

The user of the plugin should not notice that swizzling is disabled. The plugin should handle everything if everything is set up correctly. If the plugin does not handle everything, it is a bug.

**Reply by alexookah** [reply]
My app actually works just fine without `FirebaseAppDelegateProxyEnabled`.

In the documentation for firebase auth it says:
`If you prefer not to use swizzling, you can disable it by adding the flag FirebaseAppDelegateProxyEnabled to your app's Info.plist file and setting it to NO. Note that setting this flag to NO also disables swizzling for other Firebase products, including Firebase Cloud Messaging.`

I would say it's something optional because maybe someone with native code wants to have it disabled because it [overrides functions selectors](https://stackoverflow.com/a/53189296/4334709).

Back to the point. If it is disabled then the login for phone authentication **does not work**. I get the error mentioned like [above](https://github.com/capawesome-team/capacitor-firebase/discussions/254#discussioncomment-4140989).

In the documentation for firebase auth: if FirebaseAppDelegateProxyEnabled is disabled, then this needs to be done for phone-sign-in:
https://firebase.google.com/docs/auth/ios/phone-auth#appendix:-using-phone-sign-in-without-swizzling

in `didRegisterForRemoteNotificationsWithDeviceToken` the token needs to be passed in the firebase auth:
`Auth.auth().setAPNSToken(deviceToken, type: .prod)`
which I think currently is not handled, thats why I get an error.

**Comment by robingenz** [comment]
> My app actually works just fine without `FirebaseAppDelegateProxyEnabled`.

Then leave it out for now.

> In the documentation for firebase auth: if FirebaseAppDelegateProxyEnabled is disabled, then this needs to be done for phone-sign-in:
> firebase.google.com/docs/auth/ios/phone-auth#appendix:-using-phone-sign-in-without-swizzling
> 
> in didRegisterForRemoteNotificationsWithDeviceToken the token needs to be passed in the firebase auth:
> Auth.auth().setAPNSToken(deviceToken, type: .prod)
> which I think currently is not handled, thats why I get an error.

You are right. That is currently missing. Thanks for the research. I have created an issue: #256

**Reply by alexookah** [reply]
Swizzling is intended for all firebase products to not have to do manual implementation code and pass the device token between all products and analytics etc.

It should be not required to have for this awesome capacitor-firebase plugin.

I am sorry to insist so much I just want to clarify this and improve the documentation of this library :)

**Reply by robingenz** [reply]
No problem, you help me a lot to find bugs and inconsistencies.
Thank you for that and please keep going!

I used this guide, among others, during the implementation of the Firebase Messaging plugin: https://www.kodeco.com/20201639-firebase-cloud-messaging-for-ios-push-notifications
There it says:
> By default, FirebaseMessaging uses method swizzling to handle push notifications. You’ll handle all the code yourself, so turn this off using the plist entry you just added.

I need to check first what would be the consequences of enabling `FirebaseAppDelegateProxyEnabled`. Unfortunately testing all Firebase plugins is always a big effort and I will need some time.
But I will definitely check it out.

You say you have no problems with `FirebaseAppDelegateProxyEnabled` enabled. What `@capacitor-firebase` plugins do you use?

**Reply by alexookah** [reply]
I am currently using `firebase auth (phone)` and `firebase messaging`. I also tested a notification click action and did not have some issue.

I wouldn't rely on someone who just likes to create a tutorial that maybe wants to show what exactly is happening on the background from firebase. Moreover the sentence  `You’ll handle all the code yourself, so turn this off using the plist entry you just added.` is not enough that means it's required to have it disabled.

If it was mentioned in the official capacitor documentation then I would consider that it needs to be disabled.
https://capacitorjs.com/docs/guides/push-notifications-firebase

And if we agree that it is not necessary. then i would use the plugin with default behaviour and only if someone has some issue with some hybrid app or something then I would try to support the swizzling disabled option.

**Comment by robingenz** [comment]
> If it was mentioned in the official capacitor documentation then I would consider that it needs to be disabled.
capacitorjs.com/docs/guides/push-notifications-firebase

Good point!

> And if we agree that it is not necessary. then i would use the plugin with default behaviour and only if someone has some issue with some hybrid app or something then I would try to support the swizzling disabled option.

Currently I also think it is not necessary and will update the documentation. I still add the support for swizzling as it should not hurt. Finally, these plugins should provide the same functionality as the Firebase SDKs at some point.

**Reply by alexookah** [reply]
Yes exactly. But I am not sure if supporting both methods is ideal. The documentation doesn't hurt but what about the SDK code plugin?

I mean if swizzle is enabled and the code is also handling cases for both options it kinda does not make sense. I mean we either have some option like with `skipNativeAuth` called `disableSwizzle` in `capacitor.config.ts` that does handle code for swizzle disabled. 
But maintaining this is not great and also who knows if the firebase SDK is perfect that can handle both implementations logic?

Now there is already one new bug that needs to be fixed for swizzle disabled and who knows if analytics needs also some handling...e.tc.

Finally how many capacitor apps need to use swizzle disabled? It's better to have short documentation with everything a user needs rather than cover all edge cases.
Thats just my opinion.

**Comment by alexookah** [comment]
@robingenz 
I have asked the creator of the tutorial you referenced as a guide you used. Just to have another opinion in this matter.

In case you want to read: https://forums.kodeco.com/t/firebase-cloud-messaging-swizzle/180938/3

**Reply by robingenz** [reply]
Thank you for that! Good to know that there are no special reasons. In this case, we should simply take the solution with which we have fewer problems.

**Reply by alexookah** [reply]
I could try and make a PR that removes the support for SwizzlingDisabled. What do you think?

**Reply by robingenz** [reply]
I would like to keep the support as I don't think it would hurt. There are just a few lines of code missing. I was planning to do this over the weekend and then test it.
If someone then has problems with swizzling he can simply deactivate it. And everyone who has already deactivated Swizzling benefits from the fix.

**Comment by robingenz** [comment]
@alexookah **Quick update**: I disabled swizzling in my demo app and everything still seems to work. I update the docs and remove the sentence regarding `FirebaseAppDelegateProxyEnabled`.

**Comment by braincomb** [comment]
Hi @robingenz and @alexookah , I stumbled upon this discussion when searching for the same issue that I'm currently getting with iOS:

`
10.2.0 - [FirebaseMessaging][I-FCM002022] APNS device token not set before retrieving FCM Token for Sender ID 'xxxxxxxxxxx'. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.
`

I've followed all directions and instructions. I'm using the latest v1.3.0 plugin. I have APNS Auth Key uploaded to Firebase. I've tried this with and without FirebaseAppDelegateProxyEnabled.

My client side code has this piece of code only:

```ts
import { FirebaseMessaging } from '@capacitor-firebase/messaging';

const getToken = async () => {
  const result = await FirebaseMessaging.getToken();
  return result.token;
};
```

I do believe that it works, since it returns the token:

`
FirebaseMessaging getToken 3216683
⚡️  TO JS {"token":"fPGAKoBAKERlvT9L0IMhQb:APA91bHWaOgEP73T9x7eNXLIL2r5oRbFU8zv3z2_hTAU229m4gfwaQPXOcUHjfuxAONSZz65OzHkyDpRfUbR5fqro9b4fmonDnbfisngh9FYjP94U4Cy7s-DHU-pdXmQKv84NQo_g1tJ"}
`
But it doesn't feel like it actually registers it ... Since when I try to send push notifications to the iOS device, I always get the error above about APNS device token not set before retrieving FCM Token for Sender ID.

Curious if I'm missing anything that is required for this plugin to work? Any help appreciated.

**Reply by robingenz** [reply]
Unfortunately, I don't have time right now (writing my exams). I can't look at it for at least a week or two.
